syntax = "proto3";

package p2p;

service P2PService {
  rpc JoinNetwork (JoinRequest) returns (JoinResponse);
  rpc LeaveNetwork (LeaveRequest) returns (LeaveResponse);
  rpc UploadFile (FileTransferRequest) returns (FileTransferResponse);
  rpc DownloadFile (FileTransferRequest) returns (FileTransferResponse);
  rpc Ping (PingRequest) returns (PingResponse);
  rpc LookupID (LookupRequest) returns (LookupResponse);
  rpc UpdateFTable (UpdateFTableRequest) returns (UpdateFTableResponse);
  rpc GetFingerTable (Empty) returns (FingerTableResponse);
}

message JoinRequest {
  string ip = 1;
  int32 port = 2;
}

message JoinResponse {
  string predecessor_ip = 1;
  int32 predecessor_port = 2;
}

message LeaveRequest {
  string ip = 1;
  int32 port = 2;
}

message LeaveResponse {
  bool success = 1;
}

message FileTransferRequest {
  string filename = 1;
  bytes data = 2; // For upload
}

message FileTransferResponse {
  bool success = 1;
  bytes data = 2; // For download
}

message PingRequest {}

message PingResponse {
  string predecessor_ip = 1;
  int32 predecessor_port = 2;
}

message LookupRequest {
  int32 keyID = 1;
}

message LookupResponse {
  string ip = 1;
  int32 port = 2;
}

message UpdateFTableRequest {
  repeated FingerTableEntry entries = 1;
}

message UpdateFTableResponse {
  bool success = 1;
}

message FingerTableEntry {
  int32 id = 1;
  string ip = 2;
  int32 port = 3;
}

message FingerTableResponse {
  repeated FingerTableEntry entries = 1;
}

message Empty {}
